internal/cas/cas.go

```diff
@@ -0,0 +1,319 @@
+	c.cloneLock.Lock()
+	defer c.cloneLock.Unlock()
+
+	targetDir := c.prepareTargetDirectory()
+
+	// Create a temporary directory for git operations
+	tempDir, cleanup, err := c.git.CreateTempDir()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if cleanupErr := cleanup(); cleanupErr != nil {
+			(*l).Warnf("cleanup error: %v", cleanupErr)
+		}
+	}()
+
+	// Set the working directory for git operations
+	c.git.SetWorkDir(tempDir)
+
+	hash, err := c.resolveReference()
+	if err != nil {
+		return err
+	}
+
+	if !c.store.HasContent(hash) {
+		if err := c.cloneAndStoreContent(l, hash); err != nil {
+			return err
+		}
+	}
+
+	content := NewContent(c.store)
+
+	treeData, err := content.Read(hash)
+	if err != nil {
+		return err
+	}
+
+	tree, err := ParseTree(string(treeData), targetDir)
+	if err != nil {
+		return err
+	}
+
+	if err := tree.LinkTree(ctx, c.store, targetDir); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (c *CAS) prepareTargetDirectory() string {
+	targetDir := c.opts.Dir
+	if targetDir == "" {
+		targetDir = GetRepoName(c.url)
+	}
+
+	return filepath.Clean(targetDir)
+}
+
+func (c *CAS) resolveReference() (string, error) {
+	results, err := c.git.LsRemote(c.url, c.opts.Branch)
+	if err != nil {
+		return "", err
+	}
+
+	if len(results) == 0 {
+		return "", &WrappedError{
+			Op:      "clone",
+			Context: "no matching reference",
+			Err:     ErrNoMatchingReference,
+		}
+	}
+
+	return results[0].Hash, nil
+}
+
+func (c *CAS) cloneAndStoreContent(l *log.Logger, hash string) error {
+	if err := c.git.Clone(c.url, true, 1, c.opts.Branch); err != nil {
+		return err
+	}
+
+	return c.storeTreeRecursively(l, hash, "")
+}
+
+func (c *CAS) storeTreeRecursively(l *log.Logger, hash, prefix string) error {
+	tree, err := c.git.LsTree(hash, ".")
+	if err != nil {
+		return err
+	}
+
+	content := NewContent(c.store)
+	treeData := strings.Builder{}
+
+	for _, entry := range tree.Entries() {
+		fmt.Fprintf(&treeData, "%s %s %s %s\n", entry.Mode, entry.Type, entry.Hash, entry.Path)
+	}
+
+	var (
+		subTrees    []TreeEntry
+		blobEntries []TreeEntry
+	)
+
+	for _, entry := range tree.Entries() {
+		if prefix != "" {
+			entry.Path = filepath.Join(prefix, entry.Path)
+		}
+
+		switch entry.Type {
+		case "blob":
+			blobEntries = append(blobEntries, entry)
+		case "tree":
+			subTrees = append(subTrees, entry)
+		}
+	}
+
+	if err := c.storeBlobEntries(l, blobEntries); err != nil {
+		return err
+	}
+
+	for _, subTree := range subTrees {
+		if err := c.storeTreeRecursively(l, subTree.Hash, subTree.Path); err != nil {
+			return err
+		}
+	}
+
+	// We only do this if we're at the root of the repository.
+	if prefix == "" {
+		for _, file := range c.opts.IncludedGitFiles {
+			stat, err := os.Stat(filepath.Join(c.git.WorkDir, file))
+			if err != nil {
+				return err
+			}
+
+			if stat.IsDir() {
+				continue
+			}
+
+			workDirPath := filepath.Join(c.git.WorkDir, file)
+
+			data, err := os.ReadFile(workDirPath)
```

Comment:
When dealing with large files, this may potentially lead to out-of-memory errors